### PR TITLE
enhancement: Keyword Query Language (KQL) search syntax support

### PR DESCRIPTION
**Pull request by fschade**
## Description

Introduce support for [KQL](https://learn.microsoft.com/en-us/sharepoint/dev/general-development/keyword-query-language-kql-syntax-reference) syntax for search.

The functionality consists of a kql lexer and a bleve query compiler

Supported field queries:

* `Tag` search `tag:golden tag:"silver"`
* `Filename` search `name:file.txt name:"file.docx"`
* `Content` search `content:ahab content:"captain aha*"`

Supported conjunctive normal form queries:

* `Boolean` operators `AND`, `OR`, `NOT`,
* `Nesting` `(` `...SUB_QUERY...` `)`

for example:

**Query**:

`(name:"moby di*" OR tag:bestseller) AND tag:book NOT tag:read`

**Result**:

* Resources with `name: moby di*` `OR` `tag: bestseller`.
* `AND` with `tag:book`.
* `NOT` with `tag:read`.

**AST**:

```go
&ast.Ast{
  Nodes: []ast.Node{
    &ast.GroupNode{
      Base: &ast.Base{Loc: &ast.Location{...}},
      Nodes: []ast.Node{
        &ast.StringNode{...},
        &ast.OperatorNode{...},
        &ast.StringNode{...},
      },
    },
    &ast.OperatorNode{Base: &ast.Base{Loc: &ast.Location{...}}, Value: "AND"},
    &ast.StringNode{Base: &ast.Base{Loc: &ast.Location{...}}, Key: "tag", Value: "book"},
    &ast.OperatorNode{Base: &ast.Base{Loc: &ast.Location{...}}, Value: "NOT"},
    &ast.StringNode{Base: &ast.Base{Loc: &ast.Location{...}}, Key: "tag", Value: "read"},
  },
}
```

## Todos:
-  [X] `KQL-STRING-QUERY` to `AST` kql syntax subset [`KQL-TO-AST-LEXER`]
    -  [X] name (`ast.StringNode`)
    -  [X] tag (`ast.StringNode`)
    -  [X] content (`ast.StringNode`)
    -  [X] free text-keywords (`ast.StringNode`)
    -  [X] conjunctive normal form (`ast.GroupNode`)
    -  [X] operators `AND`, `OR`, `NOT` (`ast.OperatorNode`)
-  [ ] `AST` to `BLEVE-STRING-QUERY` compiler [`AST-TO-BLEVE-COMPILER`]
    -  [x] name (`ast.StringNode`)
    -  [x] tag (`ast.StringNode`)
    -  [x] content (`ast.StringNode`)
    -  [x] free text-keywords (`ast.StringNode`)
    -  [x] conjunctive normal form (`ast.GroupNode`)
    -  [ ] operators `AND`, `OR`, `NOT` (`ast.OperatorNode`)

## Related Issue
- Fixes https://github.com/owncloud/ocis/issues/7042

## How Has This Been Tested?
- unit tests

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [X] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Technical debt
- [ ] Tests only (no source changes)

## Checklist:

- [X] Code changes
- [X] Unit tests added
- [ ] Acceptance tests added
- [ ] Documentation ticket raised: <link>

### CODE REVIEW

**Comment by 2403905** [comment]
We should discuss how to handle the implicit boolean operators and NOT together. It is not explicitly described in the documentation. The rules are different for the `multiple free-text expressions`, `multiple instances of the same property restriction`, `multiple instances of the different property` and `grouping-property`.
https://github.com/owncloud/ocis/blob/46b8e437f3f46c8f50fe3b82dcf6356e9e3a473a/services/search/pkg/query/kql/factory.go#L156-L176
Should we support and how do we represent?
For example:
```
 author:"John Smith" author:"Jane Smith"
This functionally is the same as using the OR Boolean operator, as follows:
author:"John Smith" OR author:"Jane Smith
```
Is `NOT author:"John Smith" NOT author:"Jane Smith"`  became `NOT author:"John Smith" OR  NOT author:"Jane Smith` ?

```
author:"John Smith" filetype:docx
This is the same as using the AND Boolean operator, as follows:
author:"John Smith" AND filetype:docx
```
Is `NOT author:"John Smith" NOT filetype:docx"`  became `NOT author:"John Smith" AND NOT filetype:docx` ?

**Comment by fschade** [comment]
> author

the ms spec defines when to use a implicit AND and when to use a implicit OR, but negotiations aren't mentioned.

my 5 cents: since no description is available for those cases i would say, as soon as someone uses an operator as edge between two nodes:

```
┌─────────────┐                            ┌───────────────┐
│             │        ┌─────────┐         │               │
│QUERY-NODE-1 ├────────┤ OP-EDGE ├─────────┤ QUERY-NODE-2  │
│             │        └─────────┘         │               │
└─────────────┘                            └───────────────┘
```

we do not add any other implicit type there. cC.: @tbsbdr

File: services/search/Makefile [hunk 1]
```diff
@@ -24,11 +24,12 @@ docs-generate: config-docs-generate
 include ../../.make/generate.mk
 
 .PHONY: ci-go-generate
-ci-go-generate:  $(MOCKERY) # CI runs ci-node-generate automatically before this target
+ci-go-generate: $(PIGEON) $(MOCKERY) # CI runs ci-node-generate automatically before this target
 	$(MOCKERY) --dir pkg/engine --output pkg/engine/mocks --case underscore --name Engine
 	$(MOCKERY) --dir pkg/content --output pkg/content/mocks --case underscore --name Extractor
 	$(MOCKERY) --dir pkg/content --output pkg/content/mocks --case underscore --name Retriever
 	$(MOCKERY) --dir pkg/search --output pkg/search/mocks --case underscore --name Searcher
+	$(PIGEON) -o pkg/query/kql/dictionary_gen.go pkg/query/kql/dictionary.peg
```
Comment (kobergj) [inline services/search/Makefile:32]: Curious: What does this `"PIGEON"` do and why do we need it?

File: services/search/pkg/query/kql/normalize.go [hunk 2]
```diff
@@ -0,0 +1,120 @@
...
+var operatorNodeAnd = ast.OperatorNode{Base: &ast.Base{Loc: &ast.Location{Source: &implicitOperatorNodeSource}}, Value: BoolAND}
+var operatorNodeOr = ast.OperatorNode{Base: &ast.Base{Loc: &ast.Location{Source: &implicitOperatorNodeSource}}, Value: BoolOR}
+
+// NormalizeNodes Populate the implicit logical operators in the ast
+//
+// https://learn.microsoft.com/en-us/sharepoint/dev/general-development/keyword-query-language-kql-syntax-reference#constructing-free-text-queries-using-kql
+// If there are multiple free-text expressions without any operators in between them, the query behavior is the same as using the AND operator.
+// "John Smith" "Jane Smith"
+// This functionally is the same as using the OR Boolean operator, as follows:
+// "John Smith" AND "Jane Smith"
```
Comment (kobergj) [inline services/search/pkg/query/kql/normalize.go:17]: Misleading. Either a typo in line `16` (`OR`) or a typo in line `17` (`AND`)

**Review by kobergj — CHANGES_REQUESTED** [review]
Looks fluffy :+1: Just one question and one typo.

I assume documentation is done as a follow up?

File: services/search/Makefile (hunk 1 above)
Comment (fschade) [reply services/search/Makefile:32]: it converts the parsing expression grammar file to go (ast builder backbone)

**Comment by fschade** [comment]
> I assume documentation is done as a follow up?

🤣 SURE, to be more exact, KQL is not enabled at the moment, but in the follow up we will enable it and provide a readme and with the help of @mmattel a even more fluffy docs page 💋

File: services/search/pkg/query/kql/normalize.go (hunk 2 above)
Comment (fschade) [reply services/search/pkg/query/kql/normalize.go]: GOSH, what a review, you read that comment 🥇 THX!

**Review by kobergj — APPROVED** [review]

**Pull request merged**